netty/src/test/java/io/grpc/netty/ProtocolNegotiatorsTest.java

```diff
@@ -0,0 +1,213 @@
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.netty;
+
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import com.google.common.collect.Iterables;
+
+import io.grpc.netty.ProtocolNegotiators.TlsChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelPipeline;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.handler.ssl.SslHandshakeCompletionEvent;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Matchers;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+import java.util.logging.Filter;
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+import java.util.logging.Logger;
+
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLEngine;
+
+@RunWith(JUnit4.class)
+public class ProtocolNegotiatorsTest {
+  @Rule public final ExpectedException thrown = ExpectedException.none();
+  @Rule public final MockitoRule mocks = MockitoJUnit.rule();
+
+  @Mock private ChannelHandlerContext channelHandlerCtx;
+  @Mock private ChannelPipeline pipeline;
+  @Mock private ChannelHandler channelHandler;
+
+  private SSLEngine engine;
+
+  @Before
+  public void setUp() throws Exception {
+    engine = SSLContext.getDefault().createSSLEngine();
+    engine.getSession();
+  }
+
+  @Test
+  public void tlsHandler() throws Exception {
```

Comment:
Thinking about this from a practical point of view, the refactor present in this PR makes the  serverTls() essentially a one liner.  It makes sense that the test for it is equivalently small.  I don't believe this counts as cheating.  

The real test that this code works is in the interop test, but because those don't run on each PR (or at least not run by default, or whatever tests cause the coverage to run), this gives at least some confidence that this method works.  Same with the other test cases added here.  

I personally view unit tests ("small tests") as pessimistic, mainly assuming that there are bugs, and that the need to be prevented from happening.  Medium tests (aka system tests or what we call interop tests) are optimistic IMO, and assume by default that there aren't really bugs, and instead focus on making sure the correct behavior does happen.  
From this point of view, coverage feels like a unit test.  It checks that the can even possibly run, but doesn't say what it should do.  It isn't great for checking that branching patterns are hit (taking the if one time through a loop, and then the else next).  It's a necessary, but not sufficient.  

*Cheating* on coverage doesn't make sense to me.  The coverage is a feasibility report.  Will this code blow up when run?  It doesn't say whether the right or wrong thing happened.  Only the actual test code and verification does that. It cannot really inspire false confidence, because coverage is just seeing if the code *could* work.  It does a good job of this!  I think you are expecting it to be something more than it actually is, which is why you are resisting this change.  This is also why I suggested changing the OKR, because I think you are giving too much credence to coverage.